src/common.cpp

```diff
@@ -1,10 +1,93 @@
 #include "common.hh"
+#include "zarr.hh"
 
 #include "platform.h"
 
 #include <cmath>
+#include <thread>
 
 namespace common = acquire::sink::zarr::common;
+common::ThreadPool::ThreadPool(size_t n_threads,
+                               std::function<void(const std::string&)> err)
+  : contexts_{ n_threads }
+{
+    if (n_threads == 0) {
+        throw std::runtime_error("Cannot create thread pool with 0 threads.");
+    }
+
+    for (auto& ctx_ : contexts_) {
+        ctx_.should_stop = false;
+        ctx_.ready = true;
+        ctx_.thread = std::thread([this, ctx = &ctx_] { thread_worker_(ctx); });
+    }
+}
+
+common::ThreadPool::~ThreadPool() noexcept
+{
+    // spin down threads
+    for (auto& ctx : contexts_) {
+        ctx.should_stop = true;
+        ctx.cv.notify_one();
+        ctx.thread.join();
+    }
```

Comment:
Very very often, I find it necessary to pull blocking behavior like this outside of the destructor - into a `await_stop()` kind of function. Then call that at the appropriate point in the client code.

The destructor's behavior should be more severe - aborting threads without waiting for them to complete if possible.

The reason is that we want to be able to handle exceptions from the `stop()` in a controlled way, and that's hard at destruct-time.

Reply:
I'm not sure how to abort threads without just calling `thread.join()` on them, which will block. Looking around I see Windows-specific solutions. Still, good point about stopping the threads before destructing. I'll take care of that.